### DISCUSSION TITLE
Main map and base menus

### DISCUSSION THREAD

**Discussion by ramseysparrow** [induktio/thinker#61] [Ideas]
Returning to Alpha Centauri after 15 years (Thinker + PRAX now) I'm noticing a few UI aspects that I find myself missing the most, and so I thought I would pass on a few impressions and ideas. Love what Thinker and PRAX did to the game so far. While the nature of proposed improvements is much more in line with PRAX's focus than with Thinker’s, I’m posting those here seeing as PRAX is no longer actively developed. The focus lands on two core priorities that tie into one another.

**1. Recognising that keeping track of which units belong to which bases, where they are and what they are doing is the single biggest chokepoint of SMAC’s user interface on the overall.**

**2. Base Menus navigation could use a few more smart keyboard shortcuts to improve the moment-to-moment game flow.**
 
 
 
**=== 1 MAIN MAP  ===**

The first, most effective and universal method to help visualize the flow of a growing, pixelated empire and command units would be to allow a dynamic dimming and highlighting of elements on the main map to indicate the interrelations between units and their supporting bases with nothing more than holding down of a key.

While details would have to be worked out based on the engine and what’s actually possible, what I imagine is a visual filter in the style of those that PRAX demonstrated to be both possible and very effective, that only comes up for as long as a single universal control key is being held down. Let’s assume our key is [ALT]. The system works as follows: 

The filter engages on holding down [ALT] and hovering the cursor over any Base or Unit on the main map.

It causes the entire game map to dim down (either in brightness or to a single color tint, e.g. dark blue) leaving highlighted only the base and all of the units it supports. It does not matter if it is a base or a unit that is hovered over – the result is the same. 

For as long as [ALT] is held down, all highlighted units gain underneath them a description of their current actions the same way that bases normally indicate their current production, and all of their movement/patrol routes are rendered at once.

For as long as the filter is active, the right MFD minimap also changes to that normally displayed in the Base Management menu’s Support screen, effectively casting the same kind of information in the form of a mini overview.

The filter lifts the moment [ALT] key is let go.

**=== 2 BASE MENUS ===**

The second proposed QoL can exist independently of or complementarily to the first one, dealing with the Base Management menu screen rather than main map. As is in the game at present, the Base management screen allows for convenient fast switching between Bases with [left] and [right] keys, but nothing beyond that. I’d like to suggest a few extras.

**[2.A]** Base Menu’s Support screen is currently the easiest way to quickly visualize the location of both the selected Base and its supported units. I suggest shortcuts for switching between the Resource/Support/Psych sub-tabs. I’m thinking two sets to cover different keyboard sizes and preferences: 

	- holding [ctrl] while pressing the left/right arrows (main shortcut)
	- page up / page down (alternative for those using full-size pc keyboards)

**[2.B]** The Support screen is extremely useful but barely functional: it is so pixelated that things can barely be made out, cannot be zoomed in/out and doesn’t allow for highlighting of interrelations between the units it displays and those listed in the 
‘Forces Supported’ MFD. I’d try to make the Support Screen:

	- cleaner (possibly by similar type of color filtering as mentioned in point 1)
	- allow mouse scroll zooming within it (if no other ways of making it clearer)
	- allow for cross-highlighting of units on the [Support Map] and [Forces Supported] MFD by hovering over them
	- adding an up/down keyboard shortcut to cycle through ‘forces supported’ screen units without selecting them

_edited to clean up formatting_

**Comment by induktio** [comment]
There are some limitations when patching new features on the graphics especially if it requires extensive reworking on the rendering pipeline. That kind of visual filter with a dimming element for the whole map screen is probably very complicated to implement. Also supporting these kind of features while PRACX is also active is not feasible, since already Thinker has to disable some parts of the code when PRACX is running because they have to patch similar parts of the engine code.

Debugging code on two different binaries is also extra work that would not be meaningful at all, so that's why I've been thinking of transferring more of the terrain overlay/tile yield code from PRACX into Thinker. It has not been implemented yet since the code is not easily portable to another compiler and not sure really how many players would use those features. After playing some time I would think it's somewhat easy to see visually the tile yield from the map anyway. That being said, the idea about keyboard shortcuts for base window subscreens seems very useful. Next release could at least implement Ctrl+Left/Right shortcuts for those displays. Not sure about those other things yet.

**Comment by ramseysparrow** [comment]
Indeed I wrote this in anticipation that some features may be difficult or impossible to implement. I don't consider Prax critical – I only mentioned it to indicate that my suggestions are more GUI than AI oriented. So to keep confusion down, going forward we can just say we're talking here about Thinker strictly, and keep prax out of the conversation.

As I said, quickly visualising the connections between units and bases is the biggest UI chokepoint of the game, so the bottom line design challenge we're discussing is: **What can make this information available with least amount of eye squinting, clicks and moves?**

I have no coding expertise so what is possible remains entirely at your discretion, but from a pure design standpoint the quickest way I see to improve this system greatly (and assuming that the first idea I outlined is way too complex to code, which is fair enough) is to make the Support Screen somewhat interactive and, even better, work in tandem with the Forces Supported display. 

This could work in a million different ways, so I'm just spitballing ideas here. One would be to make units hovered over in Forces Supported window blink on the Support Screen. Another would be to allow clicking on units on the Support Screen directly. 
Another would be to set extra shortcuts such as [mouse 3], [mouse 4] buttons to perform the actions that are otherwise only available by right-click menu, making such things as [centre on this unit] quicker than right clicking and choosing from the menu.

I must also say that the little minimap that opens when using unit command 'go to base' or when deciding which base to attack while on comms with a friendly faction leader is very good looking and clean, giving a very good idea of what is where. Depending on what can be done by coding, I would definitely suggest that you look into that one as possible framework for showing other things clearly, such as the units positions on the Support Screen for instance. 

All food for thought really, thanks for making one of the best games ever made even better.

**Comment by induktio** [comment]
Looks like the base window support tab can be modified by patching BaseWin_draw_support function. However changing it such that it would render the area with more detail or adjust the zoom level will most likely require rewriting the function entirely. There are lots of variables that would need to be adjusted for the view to not appear glitched. So that might be looked at in later releases.

**Comment by ramseysparrow** [comment]
Glad to hear theres something there to potentially work with. 

Will look forward to future releases.

**Comment by JaggedOne** [comment]
I am using Thinker and love it.
What is PRAX and should I install it ?

**Reply by ramseysparrow** [reply]
Prax is another mod that improves GUI, allowing for things such as smooth mouse scrolling and dragging but also displaying tile resource yield, Civilisation-style, and allowing for several map filter modes to indicate such things as ownership, elevation etc via colours. 

Thinker already incorporates the main smooth scrolling and dragging features. You can install Prax to get the others. What we were discussing here was the trouble of upkeeping conpatibility between PRAX and Thinker in the future, even if it is fine now and you can use both.
